### ISSUE TITLE
Pallas TPU reduce kernel generalization gives wrong numerical results?

### ISSUE THREAD

**Issue by rdyro** [google/jax#23099]
### Description

Credit to [simveit](https://github.com/simveit) for investigating implementing a generalization of the Pallas TPU reduce kernel [here](https://jax.readthedocs.io/en/latest/pallas/tpu/pipelining.html):

The original (non-generalized kernel):
```python
def sum_kernel(x_ref, o_ref):
  @pl.when(pl.program_id(axis=0) == 0)
  def _():
    o_ref[...] = jnp.zeros_like(o_ref)

  o_ref[...] += x_ref[...]

def sum(x: jax.Array) -> jax.Array:
  grid, *out_shape = x.shape
  return pl.pallas_call(
      sum_kernel,
      grid=grid,
      # None in `block_shape` means we pick a size of 1 and squeeze it away
      in_specs=[pl.BlockSpec((None, *out_shape), lambda i: (i, 0, 0))],
      out_specs=pl.BlockSpec(out_shape, lambda i: (0, 0)),
      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
  )(x)

sum(x)
```

The user is not following the docs paragraph which say only last (minor-most) dimensions should be reduced over. However, a transposition of the user-kernel (according to what the docs suggest) also gives a numerically wrong result.

Repro code:
```python

import jax
from jax.experimental import pallas as pl
import jax.numpy as jnp
import numpy as np
import functools
from jax import random

# # Reducing over the minormost axis
#
# [https://jax.readthedocs.io/en/latest/pallas/tpu/pipelining.html](https://jax.readthedocs.io/en/latest/pallas/tpu/pipelining.html)
#
# > One last thing to note about reductions in Pallas are that they must be done
# > in the minormost (rightmost) dimensions of our grid (our grid is 1-dimensional
# > in the above example so we are reducing over its minormost dimension).
#
# > This is because the pipeline that Pallas generates using the BlockSpecs, grid
# > and kernel function does not read outputs back from HBM. Once you’ve written an
# > output value back to HBM you cannot revisit it.
#
# > Therefore, you cannot do a reduction across a grid dimension that has any
# > revisiting and therefore all reductions need to happen in the rightmost
# > dimensions.

# + id="2pzgdMNivbdq"
def sum_kernel(x_ref, o_ref):
    @pl.when(pl.program_id(axis=0) == 0)
    def _():
        o_ref[...] = jnp.zeros_like(o_ref)
    o_ref[...] += x_ref[...]

@functools.partial(jax.jit, static_argnames=["b1", "b2", "strategy", "interpret"])
def sum_fn(x: jax.Array, b1: int, b2: int, strategy: str, interpret: bool = False) -> jax.Array:
    g0, *out_shape = x.shape  # grid is the first dimension of x
    g1 = out_shape[0] // b1
    g2 = out_shape[1] // b2
    print(f"g0: {g0}, g1: {g1}, g2: {g2}")

    if strategy == "user_major_dimension":
        grid = (g0, g1, g2)
        in_specs = [pl.BlockSpec(block_shape=(None, b1, b2), index_map=lambda i, j, k: (i, j, k))]
        out_specs = pl.BlockSpec(block_shape=(b1, b2), index_map=lambda i, j, k: (j, k))
    elif strategy == "1d_minor_dimension":
        grid = (g0,)
        in_specs = [pl.BlockSpec(block_shape=(None, *out_shape), index_map=lambda k: (k, 0, 0))]
        out_specs = pl.BlockSpec(block_shape=tuple(out_shape), index_map=lambda k: (0, 0))
    else:
      grid = (g1, g2, g0)
      in_specs = [pl.BlockSpec(block_shape=(None, b1, b2), index_map=lambda i, j, k: (k, i, j))]
      out_specs = pl.BlockSpec(block_shape=(b1, b2), index_map=lambda i, j, k: (i, j))
      if strategy == "faux_3d_minor_dimension":
        assert g1 == g2 == 1, "In faux 3d minor dimension reduction, dimensions g1, g2 must == 1"
      elif strategy == "full_3d_minor_dimension":
        pass
      else:
        raise ValueError(f"{strategy = } not supported.")

    return pl.pallas_call(
        sum_kernel,
        grid=grid,
        in_specs=in_specs,
        out_specs=out_specs,
        out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype),
        interpret=interpret,
    )(x)


n = 128
# arr = jax.numpy.ones((11, 512, 512))
primes = [3, 5, 7, 11, 13]
arr = jax.numpy.ones((len(primes), n, n)) * jnp.array(primes)[:, None, None]
b1 = n // 4
b2 = n
print(arr.devices())

strategies = [
  "user_major_dimension", # user example, disregarding minor-dimension-only rule, fails
  "1d_minor_dimension",  # suggested by the docs, works
  "faux_3d_minor_dimension",  # technically equivalent to the docs, but fails
  "full_3d_minor_dimension",  # a generalization of faux_3d, docs imply OK, but fails
]

# CHOOSE THE STRATEGY HERE
strategy = strategies[1]
print(f"{strategy = }")
if strategy == "faux_3d_minor_dimension":
  b1 = n

pallas_val = sum_fn(arr, b1, b2, strategy=strategy)
ref_val = jnp.sum(arr, 0)
print(pallas_val)
print(ref_val)
np.testing.assert_array_equal(pallas_val, ref_val)
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.31
jaxlib: 0.4.31
numpy:  2.0.1
python: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
jax.devices (4 total, 4 local): [TpuDevice(id=0, process_index=0, coords=(0,0,0), core_on_chip=0) TpuDevice(id=1, process_index=0, coords=(1,0,0), core_on_chip=0) TpuDevice(id=2, process_index=0, coords=(0,1,0), core_on_chip=0) TpuDevice(id=3, process_index=0, coords=(1,1,0), core_on_chip=0)]
process_count: 1
platform: uname_result(system='Linux', node='t1v-n-8e5ad715-w-0', release='5.19.0-1022-gcp', version='#24~22.04.1-Ubuntu SMP Sun Apr 23 09:51:08 UTC 2023', machine='x86_64')
```

**Comment by dfm** [comment]
Pinging @sharadmv for visibility.

**Comment by sharadmv** [comment]
If you change the grid rank (e.g. `(g0,) -> (g1, g2, g0)`), you'll also need to change the `pl.program_id` axis in the kernel. The code works as expected when you modify the kernel to:
```
def sum_kernel(x_ref, o_ref):
    @pl.when(pl.program_id(axis=2) == 0)
    def _():
        o_ref[...] = jnp.zeros_like(o_ref)
    o_ref[...] += x_ref[...]
```

for strategies 2 and 3.

**Comment by rdyro** [comment]
That was indeed the problem! The modified example works for all strategies where the minor-most axis is reduced:
```python
def sum_kernel(x_ref, o_ref, last_axis):
    @pl.when(pl.program_id(axis=last_axis) == 0)
    def _():
        o_ref[...] = jnp.zeros_like(o_ref)
    o_ref[...] += x_ref[...]

@functools.partial(jax.jit, static_argnames=["b1", "b2", "strategy", "interpret"])
def sum_fn(x: jax.Array, b1: int, b2: int, strategy: str, interpret: bool = False) -> jax.Array:
    g0, *out_shape = x.shape  # grid is the first dimension of x
    g1 = out_shape[0] // b1
    g2 = out_shape[1] // b2
    print(f"g0: {g0}, g1: {g1}, g2: {g2}")

    if strategy == "user_major_dimension":
        grid = (g0, g1, g2)
        in_specs = [pl.BlockSpec(block_shape=(None, b1, b2), index_map=lambda i, j, k: (i, j, k))]
        out_specs = pl.BlockSpec(block_shape=(b1, b2), index_map=lambda i, j, k: (j, k))
    elif strategy == "1d_minor_dimension":
        grid = (g0,)
        in_specs = [pl.BlockSpec(block_shape=(None, *out_shape), index_map=lambda k: (k, 0, 0))]
        out_specs = pl.BlockSpec(block_shape=tuple(out_shape), index_map=lambda k: (0, 0))
    else:
      grid = (g1, g2, g0)
      in_specs = [pl.BlockSpec(block_shape=(None, b1, b2), index_map=lambda i, j, k: (k, i, j))]
      out_specs = pl.BlockSpec(block_shape=(b1, b2), index_map=lambda i, j, k: (i, j))
      if strategy == "faux_3d_minor_dimension":
        assert g1 == g2 == 1, "In faux 3d minor dimension reduction, dimensions g1, g2 must == 1"
      elif strategy == "full_3d_minor_dimension":
        pass
      else:
        raise ValueError(f"{strategy = } not supported.")

    return pl.pallas_call(
        functools.partial(sum_kernel, last_axis=len(grid) - 1),
        grid=grid,
        in_specs=in_specs,
        out_specs=out_specs,
        out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype),
        interpret=interpret,
    )(x)
```
